OpenRA.Mods.Common/Traits/Multipliers/UpgradableMultiplierTrait.cs

```diff
@@ -0,0 +1,75 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public abstract class UpgradeMultiplierTraitInfo
+	{
+		[Desc("Accepted upgrade types.")]
+		public readonly string[] UpgradeTypes = { };
+
+		[Desc("The lowest upgrade level using the scale. (Defaults to 0 if UpgradeTypes is empty.)")]
+		public readonly int BaseLevel = 1;
+
+		[Desc("Percentages to apply with the first being applied at the base level.",
+			"Repeat last entry to accept time extensions.",
+			"If no upgrade types are specified, then the first/only modifier is always applied.")]
+		public readonly int[] Modifier = { };
+
+		// Allow subclasses to specify defaults.
+		protected UpgradeMultiplierTraitInfo(int[] modifiers, string[] upgradeTypes = null, int baseLevel = 1)
+		{
+			UpgradeTypes = upgradeTypes ?? new string[0];
+			Modifier = modifiers;
+			BaseLevel = UpgradeTypes.Length > 0 ? baseLevel : 0;
+		}
+	}
+
+	public abstract class UpgradeMultiplierTrait : IUpgradable, IDisabledTrait, ISync
```

Comment:
Basing `UpgradableMultiplierTrait` would require explicit setting of `UpgradeMaxAcceptedLevel = UpgradeMinEnabledLevel + Modifier.Length`, making mod authors do unnessary counting and adding should be unacceptable and the changes needed, otherwise, to make `UpgradableMultiplierTrait` fit onto `UpgradableTrait` would clutter the code with special condition(s) to override `UpgradableTrait` such as allowing special values and virtual hooks to override the base behavior (to add that calculation dynamically) while almost all subclasses would have no need for the functionality added (`PowerMultiplier` requires notice to call `power.UpdateActor(self)` when the level changes).

It might be worthwhile to add an abstract base class with the common subset, possibly renaming `UpgradableTrait` to better describe its slightly more specialized functionality (of explicit capping).


`UpgradableMultiplierTrait` might be the reason why I first made `IUpgradable`.